proposals/p3980.md

```diff
@@ -0,0 +1,660 @@
+now that the owning library must import the `extern` declaration, there is a
+tighter association and so we reevaluated.
+
+The `extern` modifier offers a benefit for being able to verify the association
+between non-owning and owning declarations, and offers additional parity in
+modifiers. It also makes it easy for a tool to know if it's missing a
+declaration.
+
+### Only require `extern` on the first owning declaration
+
+At present, we require `extern` on _all_ owning declarations. We could instead
+only require `extern` on the first owning declaration and, if there's a separate
+forward declaration and definition, infer it for the definition. For example:
+
+```
+// `extern` on the forward declaration.
+extern class C;
+// Infer `extern` for the definition.
+class C {}
+```
+
+The decision to require `extern` on all owning declarations is based on wanting
+the forward declaration to be optional. A rule of thumb was discussed wherein if
+a forward declaration could be removed without breaking the definition (as
+defined by it being in the same lexical scope), keywords should be duplicated to
+the definition. This is not proposed as a rule because it's not clear whether
+we'll generally follow it, but it's why this particular choice is taken.
+
+### Separate require-direct-import from non-owning declarations
+
+At present, an `extern` modifier on an owning declaration serves two purposes:
+
+1.  Indicates that a non-owning `extern library` declaration _can_ exist.
+2.  Indicates the declaration must be directly imported in order to be complete.
+
+This means that:
+
+-   The presence of `extern` on an owning declaration cannot be used to
+    determine whether a non-owning declaration exists.
+    -   Because the location of a non-owning declaration isn't explicit in the
+        owning code, this may lead to a developer failing to find the non-owning
+        declaration and misunderstanding that as the non-existence of a
+        non-owning declaration.
+-   Libraries which happen to be imported by the owning declaration may freely
+    add or remove non-owning `extern library` declarations without modifying the
+    owning library.
+
+We could give distinct syntax to the two purposes, so that they could be managed
+separately. The preference at present is to use a single syntax for both
+purposes, rather than emphasizing control or correspondence.
+
+### Other `extern` syntaxes
+
+[Issue #3986](https://github.com/carbon-language/carbon-lang/issues/3986)
+discussed other syntaxes for `extern` + `extern library`. These were mainly
+`has_extern`/`is_extern`/`externed` + `extern`. The current syntax is more
+verbose, but was chosen because:
+
+-   The behavior of imports being incomplete was something leads wanted to offer
+    separate from the non-owning declaration, and `extern` was decided to be a
+    good name for that.
+-   It has extra detail in `extern library "<owning_library>"`.
```

Comment:
I feel like this isn't capturing all of the rationale from the issue here...

Working off of the text in that issue, but also putting the emphasis on different parts and trying to structure it into a bullet point list here I ended up with something more like:
```suggestion
-   Providing behavior of imports being incomplete orthogonally but using the same keyword allowed us have a single meaning for the `export` keyword and consistent usage of it on declarations.
    -   Enabling this behavior independently wasn't seen as a problem by the leads and potentially useful even if not the primary motivation. The primary motivation was orthogonality of features.
    -   Within this context `extern` seemed like an acceptable name, and alternative names seemed significantly less good. It also gives us only one new keyword.
-   Adding the owning library with `extern library "<owning_library"` will hopefully give us better diagnostics, better human readability and understandability of the code.
    -   It is _very_ verbose, but this verbosity goes on the forward declaration in the non-owning library, which we expect to be read less often.
    -   And if in practice we find the verbosity becomes a significant issue we can revisit syntaxes to address that specifically. For example, if we have significant repetiton, we might consider a grouping structure such as `extern library "..." { <many forward declarations> }`.
```

Not sure this wording works for you, and happy to have something different. But I think it would be good to incorporate some of these points:
- the specific diagnostic & understandability benefits expected
- the goal of a somewhat orthogonal decomposition of features
- the fact that `extern` wasn't endorsed as a *great* name, just that none of teh alternatives we found were better.
- the mitigating factors of the verbosity

Reply:
Thanks, this was why I asked for rationale on #3986. I've adopted this, although I'm having trouble understanding it. In particular:
 
> Providing behavior of imports being incomplete orthogonally but using the same keyword allowed us have a single meaning for the `export` keyword and consistent usage of it on declarations.

Orthogonally from what? Did you mean `extern` instead of `export`, or is there a connection to `export` I'm missing?

Reply:
> Thanks, this was why I asked for rationale on #3986. I've adopted this, although I'm having trouble understanding it. In particular:
> 
> > Providing behavior of imports being incomplete orthogonally but using the same keyword allowed us have a single meaning for the `export` keyword and consistent usage of it on declarations.
> 
> Orthogonally from what? Did you mean `extern` instead of `export`, or is there a connection to `export` I'm missing?

I meant `extern`, sorry about that.

> 
> (if you mean having `extern` on a declaration accesses two orthogonal features, creating a single meaning, wouldn't some developers use it for each feature separately -- i.e. there are three possible meanings [first feature/second/both]?)

All I meant was separating the "must import directly" and "having a forward declaration in an unowning library". We can now think about them independently, and have the latter depend on the former. ("indpendent" maybe a better word than "orthogonal" here)

Reply:
Thanks, I've tried to rephrase this, can you make sure I've captured it correctly?

Reply:
(did a couple small rephrasings to try to better mix the suggested phrasing with examples)

Reply:
Reading the new text, I feel like it conveys a very different emphasis than what I intended.

The new text seems to frame this as having access to both aspects of this as fundamentally important, whereas I was trying to suggest that this wasn't *important*, just acceptable and it let us factor the syntax in a way that ended up giving a more declarative meaning to each component, and layered semantics as you compose them.

But maybe my suggested edit wasn't clear here, maybe @zygoloid can help with an a different phrasing that is more clear?

Reply:
Trying another rephrasing, switching emphasis on what's important versus acceptable.

Note, I'm not sure if you and I just view the terms used (orthogonal, independent, declarative meaning) differently.

As implemented, I would say `extern class C;` has two possible meanings:

1. There is a `extern library "foo" class C;` declaration in a differently library.
2. This was intended to require direct imports, and shouldn't have another declaration in a different library.

If I understand correctly, you're saying that because (1) is optional, it's orthogonal and independent of (2). `extern` has multiple possible behaviors that apply, so has more declarative meaning. Am I interpreting that correctly?

For me, to try to clarify my interpretation of the words:

- (1) requires (2), so the features are related rather than orthogonal.
- The approach means using `extern` for (2) makes (1) optional, and shifts control to other libraries rather than the owning library, so I would say they're intertwined rather than independent.
- Determining whether (1) applies requires reading all imported libraries (or, clear comments). I would say the declaration has less meaning on its own because it requires more context to understand the behavior applied.

Reply:
Merging phrasing suggestion from @zygoloid

Reply:
I suggested some tweaks that Jon has applied; @chandlerc, how does it look to you now?

Reply:
Yeah, this works well for me. Thanks (and sorry it took a while to circle back here).